src/crystal/tracing.cr

```diff
@@ -0,0 +1,244 @@
+
+          System.print_error "ERROR: failed to open %s for writing: %s\n", filename, LibC.strerror(Errno.value)
+        {% end %}
+
+        LibC._exit(1)
+      end
+
+      private def self.parse_sections(bytes)
+        each_token(bytes) do |token|
+          \{% begin %}
+            case token
+            \{% for name in Section.constants %}
+              when \{{name.downcase.id.stringify}}.to_slice
+                @@sections |= Section::\{{name.id}}
+            \{% end %}
+            end
+          \{% end %}
+        end
+      end
+
+      private def self.each_token(bytes, delim = ',', &)
+        while e = bytes.index(delim.ord)
+          yield bytes[0, e]
+          bytes = bytes[(e + 1)..]
+        end
+        yield bytes[0..] unless bytes.size == 0
+      end
+
+      # :nodoc:
+      def self.log(section : String, operation : String, time : UInt64, **metadata) : Nil
+        buf = BufferIO(512).new
+        buf.write section
+        buf.write "."
+        buf.write operation
+        buf.write " "
+        buf.write time
+
+        {% unless flag?(:wasm32) %}
+          # WASM doesn't have threads (and fibers aren't implemented either)
+          #
+          # We also start to trace *before* Thread.current and other objects have
+          # been allocated, they're lazily allocated and since we trace GC.malloc we
+          # must skip the objects until they're allocated (otherwise we hit infinite
+          # recursion): malloc -> trace -> malloc -> trace -> ...
+          thread = ::Thread.current?
+
+          buf.write " thread="
+          {% if flag?(:linux) %}
+            buf.write Pointer(Void).new(thread ? thread.@system_handle : System::Thread.current_handle)
+          {% else %}
+            buf.write thread ? thread.@system_handle : System::Thread.current_handle
+          {% end %}
+          buf.write ":"
+          buf.write thread.try(&.name) || "?"
+
+          if thread && (fiber = thread.current_fiber?)
+            buf.write " fiber="
+            buf.write fiber
+          end
+        {% end %}
+
+        metadata.each do |key, value|
+          buf.write " "
+          buf.write key.to_s
+          buf.write "="
+          buf.write value
+        end
+
+        buf.write "\n"
+        # System.print_error(buf.to_slice)
+        System.print(@@handle, buf.to_slice)
+      end
+    end
+
+    # Formats and prints a log message to stderr. The generated message is
+    # limited to 512 bytes (PIPE_BUF) after which it will be truncated. Being
+    # below PIPE_BUF the message shall be written atomically to stderr,
+    # avoiding interleaved or smashed traces from multiple threads.
+    #
+    # Windows may not have the same guarantees but the buffering should limit
+    # these from happening.
+    macro trace(section, operation, tick = nil, **metadata, &block)
+      if ::Crystal::Tracing.enabled?(\{{section}})
+        %tick = \{{tick}} || ::Crystal::System::Time.ticks
```

Comment:
**question:** Is `\{{ tick }}` supposed to be nilable at runtime? If not, the runtime `||` could be avoided:
```suggestion
        %tick = \{{tick || "::Crystal::System::Time.ticks".id }}
```
(Or set the default parameter value directly to `"::Crystal::System::Time.ticks".id`.)